daemon/containerd/image_snapshot.go

```diff
@@ -63,22 +61,151 @@ func (i *ImageService) PrepareSnapshot(ctx context.Context, id string, parentIma
 	ls := i.client.LeasesService()
 	lease, err := ls.Create(ctx, leases.WithID(id))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	ctx = leases.WithLease(ctx, lease.ID)
 
-	snapshotter := i.client.SnapshotService(i.StorageDriver())
+	if err := i.prepareInitLayer(ctx, id, parentSnapshot, initFunc); err != nil {
+		return nil, err
+	}
 
-	if err := i.prepareInitLayer(ctx, id, parentSnapshot, setupInit); err != nil {
-		return err
+	sn := i.client.SnapshotService(i.StorageDriver())
+	layer := &rwLayer{
+		id:              id,
+		snapshotterName: i.StorageDriver(),
+		snapshotter:     sn,
+		refCountMounter: i.refCountMounter,
+		lease:           lease,
 	}
 
 	if !i.idMapping.Empty() {
-		return i.remapSnapshot(ctx, snapshotter, id, id+"-init")
+		return layer, i.remapSnapshot(ctx, sn, id, id+"-init")
+	}
+
+	_, err = sn.Prepare(ctx, id, id+"-init")
+	return layer, err
```

Comment:
Is this expected to return both `rwlayer` on error? Or should this return `nil` ? Perhaps


```go
_, err = sn.Prepare(ctx, id, id+"-init")
if err != nil {
	return nil, err
}
return layer, err
```


We don't need to keep track of the mounts returned by `sn.Prepare()` anywhere?

Reply:
Done.
No, we don't need to keep track of the mounts here.